Задание 1. Смоделировать игру против лотерейного автомата типа «777». 


In [35]:
import numpy as np

def play_game_numpy():
    # Игрок платит 1 рубль
    player_money = -1

    # Получаем случайное целое число от 0 до 999
    result = np.random.randint(0, 1000)

    # Определяем выигрыш игрока
    if result == 777:
        player_money += 200
    elif result == 999:
        player_money += 100
    elif result == 555:
        player_money += 50
    elif result == 333:
        player_money += 15
    elif result == 111:
        player_money += 10
    elif result % 100 == 77:
        player_money += 5
    elif result % 10 == 7:
        player_money += 3
    elif result % 100 == 0:
        player_money += 2
    elif result % 10 == 0:
        player_money += 1

    return player_money

# Проведем несколько 10 тысяч игр и посчитаем средний выигрыш/проигрыш
num_games = 10000
total_money = np.sum([play_game_numpy() for _ in range(num_games)])
average_money = total_money / num_games

print(f"Средний выигрыш/проигрыш за одну игру: {average_money} руб.")

# Выводим информацию о том, выгодна ли игра для игрока
if average_money > 0:
    print("Игра выгодна для игрока.")
else:
    print("Игра не выгодна для игрока.")

Средний выигрыш/проигрыш за одну игру: -0.1755 руб.
Игра не выгодна для игрока.


Задание 2. Теоретически рассчитать средний выигрыш (проигрыш) и сравнить с результатами моделирования.

In [36]:
p_777 = 1 / 1000  # Вероятность выпадения 777
p_999 = 1 / 1000  # Вероятность выпадения 999
p_555 = 1 / 1000  # Вероятность выпадения 555
p_333 = 1 / 1000  # Вероятность выпадения 333
p_111 = 1 / 1000  # Вероятность выпадения 111
p_x77 = 9 / 1000  # Вероятность выпадения *77 (где x - любая цифра)
p_x7 = 90 / 1000  # Вероятность выпадения **7 (где x - любая цифра)
p_x00 = 10 / 1000  # Вероятность выпадения *00 (где x - любая цифра)
p_x0 = 90 / 1000  # Вероятность выпадения **0 (где x - любая цифра)

average_gain_or_loss_theoretical = (
    p_777 * 200 +
    p_999 * 100 +
    p_555 * 50 +
    p_333 * 15 +
    p_111 * 10 +
    p_x77 * 5 +
    p_x7 * 3 +
    p_x00 * 2 +
    p_x0 * 1 -
    1  # Стоимость каждой игры
)

print(f"Теоретический средний выигрыш (проигрыш): {average_gain_or_loss_theoretical} руб.")

Теоретический средний выигрыш (проигрыш): -0.19999999999999996 руб.


Выполним проверку. Будет ли стремиться моделирование игры с помощью метода Монте-Карло к теоретическому рассчету при увеличении количества проверок с 10 тысяч до 1 миллиона.

In [34]:
import numpy as np

def play_game_numpy():
    # Игрок платит 1 рубль
    player_money = -1

    # Получаем случайное целое число от 0 до 999
    result = np.random.randint(0, 1000)

    # Определяем выигрыш игрока
    if result == 777:
        player_money += 200
    elif result == 999:
        player_money += 100
    elif result == 555:
        player_money += 50
    elif result == 333:
        player_money += 15
    elif result == 111:
        player_money += 10
    elif result % 100 == 77:
        player_money += 5
    elif result % 10 == 7:
        player_money += 3
    elif result % 100 == 0:
        player_money += 2
    elif result % 10 == 0:
        player_money += 1

    return player_money

# Проведем  миллион игр и посчитаем средний выигрыш/проигрыш
num_games = 1000000
total_money = np.sum([play_game_numpy() for _ in range(num_games)])
average_money = total_money / num_games

print(f"Средний выигрыш/проигрыш за одну игру: {average_money} руб.")

# Выводим информацию о том, выгодна ли игра для игрока
if average_money > 0:
    print("Игра выгодна для игрока.")
else:
    print("Игра не выгодна для игрока.")

Средний выигрыш/проигрыш за одну игру: -0.197526 руб.
Игра не выгодна для игрока.


Проверка пройдена успешно Теоритеческий проигрыш = -0.199 руб.. При 10 тысяч игр смоделированный выигрыш = -0.1755 руб, а при миллионе игр = -0.197526 руб.